BaikalDataGenerator/Source/render.h

```diff
@@ -51,120 +60,54 @@ class Render
-             class... Args1,
-             class... Args2>
-    void GenerateDataset(const TCamStatesRange<CameraInfo, Args1 ...>& cam_states,
-                         const TLightsRange<LightInfo, Args2 ...>& lights,
-                         const std::filesystem::path& lights_dir,
-                         const std::vector<size_t>& spp,
-                         const std::filesystem::path& output_dir,
-                         bool gamma_correction_enabled = true,
-                         size_t start_cam_id = 0);
+           std::uint32_t num_bounces,
+           unsigned device_idx);
+
+    void AttachLight(LightObject* light);
+
+    /// This function generates dataset for network training
+    ///
+    /// @param output_dir Output directory to save dataset
+    /// @param scene_name Scene name
+    /// @param cameras_start_idx The save index of the 1st camera
+    /// @param cameras_end_idx The save index of the last camera
+    /// @param cam_states Camera states range
+    /// @param lights - lights range
+    /// 'spp' - spp vector
+    /// 'gamma_correction_enabled' - flag to enable/disable gamma correction
+    /// 'start_cam_id' - the number starting from will be named generated samples
+
+    void SaveMetadata(const std::filesystem::path& output_dir,
+                      const std::string& scene_name,
+                      unsigned cameras_start_idx,
+                      unsigned cameras_end_idx,
+                      int cameras_index_offset,
+                      bool gamma_correction_enabled) const;
+
+    void GenerateSample(CameraObject* camera,
+                        int camera_idx,
+                        const std::vector<size_t>& spp,
+                        const std::filesystem::path& output_dir,
+                        bool gamma_correction_enabled);
 
     ~Render();
 
 private:
-    void UpdateCameraSettings(const CameraInfo& cam_state);
-
-    void SetLight(const LightInfo& light, const std::filesystem::path& lights_dir);
-
-    template<template<class, class...> class TCamStatesRange, class... Args>
-    void SetLightConfig(const TCamStatesRange<LightInfo, Args...>& lights,
-                        const std::filesystem::path& lights_dir)
-    {
-        for (const auto& light : lights)
-        {
-            SetLight(light, lights_dir);
-        }
-    }
-
     void SaveOutput(const OutputInfo& info,
                     const std::string& name,
                     bool gamma_correction_enabled,
                     const std::filesystem::path& output_dir);
 
-    void GenerateSample(const CameraInfo& cam_state,
-                        const std::vector<size_t>& spp,
-                        const std::filesystem::path& output_dir,
-                        bool gamma_correction_enabled,
-                        size_t start_cam_id);
-
-    void SaveMetadata(const std::filesystem::path& output_dir) const;
-
-    std::filesystem::path m_scene_file;
+    SceneObject* m_scene = nullptr;
```

Comment:
Why not unique_ptr. Yes, I know about the problem with destructor, but we alreadyhave workaround.

Reply:
or shared_ptr

Reply:
The render does not own the scene, a client does. A scene is passed through the shared object boundary as `rpr_scene` (= `void*`), it's not a good idea to wrap an extern pointer into a smart pointer here.